Add a plugin to provide autoimport functionality

bagel897 opened the pull request:
# Autoimport plugin for pylsp
Depends on accompanying pull request in rope
@bagel897  
Closes #34
## Prereqs
Rope > 1.0.0 (AKA git builds rn)  
## Demo Images
![image](https://user-images.githubusercontent.com/57874654/165593184-98d44550-cf6f-4670-934d-1df452c5dc36.png)
can detect external modules
![image](https://user-images.githubusercontent.com/57874654/165593232-137b5491-4ad9-4854-8057-2fd93368290f.png)
grabs names from the external modules
![image](https://user-images.githubusercontent.com/57874654/165593396-cc2743d3-dd43-4047-a83c-4ec0177c6987.png)

## Features
- [x] Score-based sorting algorithm with thresholding. Picks the 25 best options.
Will take suggestions on how to improve the basic algorithm
- [x] Relatively high preformance using sqllite3 and ast.
Besides the cache generation, it preforms fine on my machine
- [x] Avoids suggestions on single line comments, from/normal imports, function/classes, dots
Most of the test cases are to cover this and it should be pretty functional.
- [x] Most completion related tests should be in rope
the main ones are in parsing existing names and whether to suggest imports
- [x] inserts all imports at the bottom of the set of imports. Recommendation is to use the isort extension.
- [x] Watches for changes (but not deleted files) and regenerates cache for saved files.
### Won't fix
1. Correct detection of type of statements from  ``__init__.py`` modules. Would be difficult to preform type inference on.
![image](https://user-images.githubusercontent.com/57874654/163751826-6663682d-ddea-41ef-88cc-dc2a6f3215fc.png)
2. Doesn't handle deleted files but this is a relatively rare.
3. Without a progress indicator, the user won't know when its ready (201) I have a WIP implementation based on #236 
4. Doesn't avoid suggestions on multiple line comments/strings, can fix if deemed important

bagel897 committed:
initial autoimport work

bagel897 committed:
provide suggestions

bagel897 committed:
use str for sorting

bagel897 committed:
textEdit to actually insert edits

bagel897 committed:
use parso to decide to use autoimport

bagel897 committed:
use fixture on test suite, use new search_full api, ignore statments already in document

bagel897 committed:
ignore class, dots, import statements

bagel897 committed:
use thresholding for sorting

bagel897 committed:
implement document_did_save, adjust sorting

bagel897 committed:
update docs, place imports correctly.

bagel897 committed:
update to use sqlite implementation

bagel897 committed:
Merge branch 'develop' of https://github.com/python-lsp/python-lsp-server into autoimport

rchl commented on pylsp/plugins/rope_autoimport.py:
```diff
@@ -0,0 +1,240 @@
+
+@hookimpl
+def pylsp_settings():
+    # Default rope_completion to disabled
+    return {"plugins": {"rope_autoimport": {"enabled": True}}}
```
Should be disabled by default probably? Both here and in the schema and documentation.

bagel897 replied:
Done

bagel897 committed:
clean up, bump rope to 1.1.1, make default disabled

rchl commented on CONFIGURATION.md:
```diff
@@ -59,7 +59,7 @@ This server can be configured using `workspace/didChangeConfiguration` method. E
 | `pylsp.plugins.pylint.executable` | `string` | Executable to run pylint with. Enabling this will run pylint on unsaved files via stdin. Can slow down workflow. Only works with python3. | `null` |
 | `pylsp.plugins.rope_completion.enabled` | `boolean` | Enable or disable the plugin. | `false` |
-| `pylsp.plugins.rope_autoimport.enabled` | `boolean` | Enable or disable the plugin. | `true` |
+| `pylsp.plugins.rope_autoimport.enabled` | `boolean` | Enable or disable the plugin. | `false` |
```
if those are supposed to be in sorted order then should be flipped

bagel897 committed:
fix: schema order

bagel897 committed:
redo test suite

bagel897 committed:
use type hint

bagel897 committed:
Move autoimport object into workspace.

Other smaller changes:
 - Allow memory only database for testing.
 - Configuration parameter for memory database

bagel897 committed:
format

bagel897 committed:
fix closing issues

bagel897 commented:
so this is almost ready:
1. It needs a fix for windows in rope https://github.com/python-rope/rope/pull/477 and associated release
2. Its still a little rough around where to provide suggestions and slow when it does. Ideally, some of this can be used in jedi and rope completions.
3. It generates a large cache which takes a while. Normally, this will be on-disk and I will work on improving it. But in CI, this takes a while and must be done seperately for each test and kept in memory. Either I change the scope of the associated fixtures to a module scope, or we have 2m long testing runs. In the future, I'll make it use configured dependencies automatically.

bagel897 committed:
Update pyproject.toml

bagel897 committed:
Merge branch 'develop' into autoimport

bagel897 commented:
Some notes about the future of this. I'm working on splitting it out to its own repository where 1-3 things will happen.
1. Rope-autoimport will use autoimport-core under the hood
2. Pylsp will use autoimport-core directly
3. Jedi will use autoimport-core, negating the requirements for any changes in pylsp.

But for now, this is perfectly functional code

bagel897 committed:
Merge branch 'develop' of https://github.com/python-lsp/python-lsp-server into autoimport

bagel897 committed:
fix: config

bagel897 committed:
fix: respect memory preference

ccordoba12 commented:
Hey @bageljrkhanofemus, thanks for your work on this! It looks really promising.

About what you said:

> I'm working on splitting it out to its own repository

We can invite you to our organization so you can create a repo for your plugin here and continue your work there. I'd say that's the best course of action.

bagel897 commented:
The plugin will stay here, but I'm trying to split out the underlying [library](https://github.com/bageljrkhanofemus/autoimport-core ). While it was developed *for* pylsp, I don't see why I'd need to put it on the pylsp organization.

ccordoba12 commented:
> While it was developed for pylsp, I don't see why I'd need to put it on the pylsp organization.

Some plugin authors prefer to maintain them in separate repos and I understood you were thinking on doing that.

bagel897 commented:
I can't because I must store the AutoImport object in the workspace (to persist in-memory databases), and normal plugins cannot do that. Ideally, that becomes a hook so they can. I also think having to install separate plugins is bad UX

ccordoba12 commented:
> I also think having to install separate plugins is bad UX

I understand, but the problem with all plugins being part of the core is that after they are merged, it's another thing that maintainers need to take care of, which can become overwhelming quite quickly.

bagel897 commented:
Fair, but unless there's a way to persist objects into workspace from plugins autoimport still needs to be here.
The biggest issue is that I wrote parso parsers in the plugin, but I don't know a better solution. I can bundle them into autoimport-core but we'd still be doing the AST parsing twice. Ideally, Jedi could handle that but that has its own issues.

ccordoba12 commented:
> Fair, but unless there's a way to persist objects into workspace from plugins autoimport still needs to be here.

Ok, no problem.

bagel897 committed:
Merge branch 'develop' of https://github.com/python-lsp/python-lsp-server into autoimport

bagel897 committed:
fix: pylint errors

bagel897 committed:
Make test data persist

wudstrand commented:
What's the status on this?

bagel897 commented:
> What's the status on this?


This PR needs to be reviewed. Possible concerns:
1. The tests timeout because they all index separately in memory - its not a cheap process to do repeatedly, but is a testing only issue. I have some ideas to fix that - namely dependency detection or making the autoimport object session scoped.
2. There's alot of parso code that I'd imagine the pylsp team understandably doesn't want to maintain here but I cannot put into rope
3. Progress support is in a separate PR. If that gets merged first, then I need to update this one or make a second PR.

I haven't heard that any are a blocker from the pylsp team so I can't really answer more.

ccordoba12 commented:
> The tests timeout because they all index separately in memory

You can skip them in Mac if they fail too much on that platform.

> There's alot of parso code that I'd imagine the pylsp team understandably doesn't want to maintain here but I cannot put into rope

That's fine as long as you commit to maintain it in the future.

> Progress support is in a separate PR

Does that PR need be merged too to finish your work to add auto-import support?

bagel897 committed:
Merge branch 'develop' of https://github.com/python-lsp/python-lsp-server into autoimport

bagel897 committed:
Switch to jedi get_names

bagel897 committed:
tests: use session scoped workspace

bagel897 committed:
fix pylint errors

bagel897 committed:
Use MAX_SIZE, don't use tuple unpacking

bagel897 commented:
- [x] Mac testing times - I made autoimport a session scoped fixture to reduce test timings. This should fix it for macs
- [x] Parsers - I think that the code I wrote should be similar to jedi's own code - I'm investigating if the Jedi API can replace some of my code. I've replaced the easier parser (get_names). The harder one (should_insert) is more specialized but jedi theoretically implements it.
- [x] Progress support - I have a second branch which I'll rebase when both PRs are merged, unless the progress PR gets merged first.

bagel897 commented:
> > There's alot of parso code that I'd imagine the pylsp team understandably doesn't want to maintain here but I cannot put into rope
> 
> That's fine as long as you commit to maintain it in the future.
 
As a student, I cannot reliably do so. I'll try to fix bugs but I make no guarantees on response time since academics have more precedence. You may find the same api in Jedi, but I couldn't find it. Rope cannot handle this since its ast based and it needs error-recovery.

ccordoba12 commented:
> As a student, I cannot reliably do so. I'll try to fix bugs but I make no guarantees on response time since academics have more precedence

Ok, I understand and no problem with that. The most important thing for us is that you're willing to address bugs related to this new functionality. The time you need to respond to them is up to you, of course.

bagel897 commented:
Yes, but its probably better to mark it off by default, and explain that it is experimental unless it is replaced in jedi/rope

ccordoba12 commented:
Agreed, that's fine for me too.

mostley commented:
Can I help somehow to make this happen?

bagel897 commented:
> Can I help somehow to make this happen?

If Jedi manifested a should_insert API, that'd help - but isn't necessary for the merge. I'm not sure what else is needed.

ccordoba12 commented:
I didn't know this was ready. I'll try to review it this week so we can include it in our next version.

ccordoba12 commented on docs/autoimport.md:
```diff
@@ -0,0 +1,18 @@
+# Autoimport for pylsp
+Requirements:
+1. install ``python-lsp-server[rope]``
+2. set ``pylsp.plugins.rope_autoimport.enabled`` to ``true`` 
+## Startup
+Autoimport will generate an autoimport sqllite3 database in .ropefolder/autoimport.db on startup.  
+This will take a few seconds but should be much quicker on future runs.
+## Usage 
+Autoimport will provide suggestions to import names from everything in ``sys.path``. You can change this by changing where pylsp is running or by setting rope's 'python_path' option. 
+It will suggest modules, submodules, keywords, functions, and classes.  
+
+Since autoimport inserts everything towards the end of the import group, its recommended you use the isort [plugin](https://github.com/paradoxxxzero/pyls-isort).
+
+## Credits
+ - Most of the code was written by me, @bageljrkhanofemus
```
```suggestion
# Autoimport for pylsp

Requirements:
1. Install ``python-lsp-server[rope]``
2. Set ``pylsp.plugins.rope_autoimport.enabled`` to ``true`` 

## Startup

Autoimport will generate an autoimport sqllite3 database in .ropefolder/autoimport.db on startup.  
This will take a few seconds but should be much quicker on future runs.

## Usage 

Autoimport will provide suggestions to import names from everything in `sys.path`. You can change this by changing where pylsp is running or by setting rope's `python_path` option. 
It will suggest modules, submodules, keywords, functions, and classes.  

Since autoimport inserts everything towards the end of the import group, its recommended you use the isort [plugin](https://github.com/paradoxxxzero/pyls-isort).

## Credits
 - Most of the code was written by me, @bagel897 
```

bagel897 replied:
Fixed in f5a2992d57c96e66191a04691f7947b01e28af27 (just used prettier)

ccordoba12 commented on pylsp/config/schema.json:
```diff
@@ -353,6 +353,14 @@
       "type": ["string", "null"],
       "default": null,
       "description": "Executable to run pylint with. Enabling this will run pylint on unsaved files via stdin. Can slow down workflow. Only works with python3."
+    },"pylsp.plugins.rope_autoimport.enabled": {
+      "type": "boolean",
+      "default": false,
+      "description": "Enable or disable the plugin."
+    },"pylsp.plugins.rope_autoimport.memory": {
```
```suggestion
    },
    "pylsp.plugins.rope_autoimport.enabled": {
      "type": "boolean",
      "default": false,
      "description": "Enable or disable autoimport."
    },
    "pylsp.plugins.rope_autoimport.memory": {
```

Please don't forget to update CONFIGURATION.md after this.

ccordoba12 commented on pylsp/plugins/rope_autoimport.py:
```diff
@@ -0,0 +1,211 @@
+import logging
```
```suggestion
# Copyright 2022- Python Language Server Contributors.

import logging
```

bagel897 replied:
Fixed in 76f68de3a9b56c9f3ee014bee0da0bd141f8547a

ccordoba12 commented on pylsp/plugins/rope_autoimport.py:
```diff
@@ -0,0 +1,211 @@
+            },
+            "detail": _document(suggestion.import_statement),
+            "additionalTextEdits": [edit],
+
```
```suggestion
```

Unnecessary blank line.

bagel897 replied:
7f40c74

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -76,7 +91,8 @@ def root_uri(self):
 
     def is_local(self):
-        return (self._root_uri_scheme in ['', 'file']) and os.path.exists(self._root_path)
+        return (self._root_uri_scheme in ['', 'file']) and os.path.exists(
+            self._root_path)
```
- Please revert this change because the line length limit of this project is 120 columns.
- Also, please refrain of making this kind of changes in the future because they don't add anything significant to your work and make PRs harder to review.

bagel897 replied:
f5a2992d57c96e66191a04691f7947b01e28af27 (sorry for running black)

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -89,7 +105,9 @@ def get_maybe_document(self, doc_uri):
 
     def put_document(self, doc_uri, source, version=None):
-        self._docs[doc_uri] = self._create_document(doc_uri, source=source, version=version)
+        self._docs[doc_uri] = self._create_document(doc_uri,
+                                                    source=source,
+                                                    version=version)
```
Please revert.

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -107,15 +125,25 @@ def apply_edit(self, edit):
 
     def show_message(self, message, msg_type=lsp.MessageType.Info):
-        self._endpoint.notify(self.M_SHOW_MESSAGE, params={'type': msg_type, 'message': message})
+        self._endpoint.notify(self.M_SHOW_MESSAGE,
+                              params={
+                                  'type': msg_type,
+                                  'message': message
+                              })
```
Please revert.

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -107,15 +125,25 @@ def apply_edit(self, edit):
         """Return the source roots for the given document."""
-        files = _utils.find_parents(self._root_path, document_path, ['setup.py', 'pyproject.toml']) or []
-        return list({os.path.dirname(project_file) for project_file in files}) or [self._root_path]
+        files = _utils.find_parents(self._root_path, document_path,
+                                    ['setup.py', 'pyproject.toml']) or []
+        return list({os.path.dirname(project_file)
+                     for project_file in files}) or [self._root_path]
```
Please revert.

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -107,15 +125,25 @@ def apply_edit(self, edit):
 
     def publish_diagnostics(self, doc_uri, diagnostics):
-        self._endpoint.notify(self.M_PUBLISH_DIAGNOSTICS, params={'uri': doc_uri, 'diagnostics': diagnostics})
+        self._endpoint.notify(self.M_PUBLISH_DIAGNOSTICS,
+                              params={
+                                  'uri': doc_uri,
+                                  'diagnostics': diagnostics
+                              })
```
Please revert.

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -128,10 +156,20 @@ def _create_document(self, doc_uri, source=None, version=None):
 class Document:
 
-    def __init__(self, uri, workspace, source=None, version=None, local=True, extra_sys_path=None,
+    def __init__(self,
+                 uri,
+                 workspace,
+                 source=None,
+                 version=None,
+                 local=True,
+                 extra_sys_path=None,
```
Please revert.

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -154,7 +192,8 @@ def __str__(self):
         # pylint: disable=import-outside-toplevel
         from rope.base import libutils
-        return libutils.path_to_resource(self._rope_project_builder(rope_config), self.path)
+        return libutils.path_to_resource(
+            self._rope_project_builder(rope_config), self.path)
```
Please revert.

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -218,7 +257,8 @@ def apply_change(self, change):
     def offset_at_position(self, position):
         """Return the byte-offset pointed at by the given position."""
-        return position['character'] + len(''.join(self.lines[:position['line']]))
+        return position['character'] + len(''.join(
+            self.lines[:position['line']]))
```
Please revert.

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -241,7 +281,8 @@ def word_at_position(self, position):
     def jedi_names(self, all_scopes=False, definitions=True, references=False):
         script = self.jedi_script()
-        return script.get_names(all_scopes=all_scopes, definitions=definitions,
+        return script.get_names(all_scopes=all_scopes,
+                                definitions=definitions,
```
Please revert.

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -251,7 +292,8 @@ def jedi_script(self, position=None, use_document_path=False):
 
         if self._config:
-            jedi_settings = self._config.plugin_settings('jedi', document_path=self.path)
+            jedi_settings = self._config.plugin_settings(
+                'jedi', document_path=self.path)
```
Please revert.

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -262,8 +304,10 @@ def jedi_script(self, position=None, use_document_path=False):
 
-        environment = self.get_enviroment(environment_path, env_vars=env_vars) if environment_path else None
-        sys_path = self.sys_path(environment_path, env_vars=env_vars) + extra_paths
+        environment = self.get_enviroment(
+            environment_path, env_vars=env_vars) if environment_path else None
+        sys_path = self.sys_path(environment_path,
+                                 env_vars=env_vars) + extra_paths
```
Please revert.

ccordoba12 commented on pylsp/workspace.py:
```diff
@@ -302,6 +345,7 @@ def sys_path(self, environment_path=None, env_vars=None):
         # TODO: when safe to break API, use env_vars explicitly to pass to create_environment
         path = list(self._extra_sys_path)
-        environment = self.get_enviroment(environment_path=environment_path, env_vars=env_vars)
+        environment = self.get_enviroment(environment_path=environment_path,
+                                          env_vars=env_vars)
```
Please revert.

ccordoba12 commented on pyproject.toml:
```diff
@@ -33,7 +33,7 @@ all = [
     "pyflakes>=2.4.0,<2.5.0",
     "pylint>=2.5.0",
-    "rope>=0.10.5",
+    "rope>=1.1.1",
```
```suggestion
    "rope>1.2.0",
```

To match the requirement below.

bagel897 replied:
Fixed in a8814981c6447dc6b352bbbb05db18c3df82aeb0

ccordoba12 commented on test/plugins/test_autoimport.py:
```diff
@@ -0,0 +1,236 @@
+from typing import Dict, List
```
```suggestion
# Copyright 2022- Python Language Server Contributors.

from typing import Dict, List
```

bagel897 replied:
Fixed in 76f68de3a9b56c9f3ee014bee0da0bd141f8547a

ccordoba12 commented on test/plugins/test_autoimport.py:
```diff
@@ -0,0 +1,236 @@
+#     workspace.rm_document(DOC_URI)
+
+
+class test_should_insert:
```
```suggestion
class TestShouldInsert:
```

We prefer snake-cased names for classes.

ccordoba12 reviewed (commented):
Thanks a lot @bagel897 for your work on this! I left a very lightweight review for you, mostly about style issues. Otherwise looks good to me.

bagel897 committed:
Use snake-cased name

Co-authored-by: Carlos Cordoba <ccordoba12@gmail.com>

bagel897 committed:
Update pylsp/config/schema.json

Co-authored-by: Carlos Cordoba <ccordoba12@gmail.com>

bagel897 committed:
Formatting changes

bagel897 committed:
Add copyright headers

bagel897 committed:
Merge branch 'autoimport' of github.com:bageljrkhanofemus/python-lsp-server into autoimport

bagel897 committed:
Restore Optional Import

bagel897 committed:
update configuration

bagel897 committed:
Update dep names

bagel897 committed:
style: remove extra line

bagel897 committed:
fix: single . handling

bagel897 committed:
style: reformat file

bagel897 committed:
Fix another line length issue

bagel897 commented:
Pushed a bugfix and the style changes. Let me know if there's anything else (or if the CI fails, I'll check). You might want to squash the PR, since its a lot of poorly labeled commits, but up to you. Thanks for the feedback @ccordoba12!

ccordoba12 commented on pylsp/plugins/rope_autoimport.py:
```diff
@@ -0,0 +1,219 @@
+    if first_child == word_node:
+        return True  # If the word is the first word then its fine
+    if len(expr.children) > 1:
+        if any(node.type == "operator" and "." in node.value
+               or node.type == "trailer" for node in expr.children):
```
```suggestion
        if any(node.type == "operator" and "." in node.value or
               node.type == "trailer" for node in expr.children):
```

I think this should solve the style issue flagged by our CIs.

ccordoba12 reviewed (commented):
Thanks for addressing my previous review @bagel897! One last suggestion for you then this should be ready.

ccordoba12 commented:
> You might want to squash the PR, since its a lot of poorly labeled commits, but up to you

Don't worry about that, we squash merge all PRs in this repo.

bagel897 committed:
Fix style issue in pylsp/plugins/rope_autoimport.py

Co-authored-by: Carlos Cordoba <ccordoba12@gmail.com>

ccordoba12 reviewed (approved):
Thanks @bagel897 for this addition! It's a great improvement!!
